Documentation/Proposals/NNNN-custom-test-execution-traits.md

```diff
@@ -0,0 +1,376 @@
+  ///   - testCase: The test case, if any, under which `function` is being
+  ///     performed. This is `nil` when invoked on a suite.
+  ///
+  /// - Throws: Whatever is thrown by `function`, or an error preventing
+  ///   execution from running correctly.
+  ///
+  /// This function is called for each ``Trait`` on a test suite or test
+  /// function which has a non-`nil` value for ``Trait/customTestExecutor-1dwpt``.
+  /// It allows additional work to be performed before or after the test runs.
+  ///
+  /// This function is invoked once for the test its associated trait is applied
+  /// to, and then once for each test case in that test, if applicable. If a
+  /// test is skipped, this function is not invoked for that test or its cases.
+  ///
+  /// Issues recorded by this function are associated with `test`.
+  func execute(_ function: @Sendable () async throws -> Void, for test: Test, testCase: Test.Case?) async throws
+}
+
+public protocol Trait: Sendable {
+  // ...
+
+  /// The type of the custom test executor for this trait.
+  ///
+  /// The default type is `Never`.
+  associatedtype CustomTestExecutor: CustomTestExecuting = Never
+
+  /// The custom test executor for this trait, if any.
+  ///
+  /// If this trait's type conforms to ``CustomTestExecuting``, the default
+  /// value of this property is `self` and this trait will be used to customize
+  /// test execution. This is the most straightforward way to implement a trait
+  /// which customizes the execution of tests.
+  ///
+  /// However, if the value of this property is an instance of another type
+  /// conforming to ``CustomTestExecuting``, that instance will be used to
+  /// perform custom test execution instead.  Otherwise, the default value of
+  /// this property is `nil` (with the default type `Never?`), meaning that
+  /// custom test execution will not be performed for tests this trait is
+  /// applied to.
+  var customTestExecutor: CustomTestExecutor? { get }
+}
+
+extension Trait {
+  // ...
+
+  // The default implementation, which returns `nil`.
+  public var customTestExecutor: CustomTestExecutor? { get }
+}
+
+extension Trait where CustomTestExecutor == Self {
+  // Returns `self`.
+  public var customTestExecutor: CustomTestExecutor? { get }
+}
+
+extension Never: CustomTestExecuting {
+  public func execute(_ function: @Sendable () async throws -> Void, for test: Test, testCase: Test.Case?) async throws
+}
+```
+
+Here is a complete example of the usage scenario described earlier, showcasing
+the proposed APIs:
+
+```swift
+@Test(.mockAPICredentials)
+func example() {
+  // ...validate API usage, referencing `APICredentials.current`...
+}
+
+struct MockAPICredentialsTrait: TestTrait, CustomTestExecuting {
+  func execute(_ function: @Sendable () async throws -> Void, for test: Test, testCase: Test.Case?) async throws {
+    let mockCredentials = APICredentials(apiKey: "...")
+    try await APICredentials.$current.withValue(mockCredentials) {
```

Comment:
We think task locals are going to be a particularly common use case, right? Perhaps we should build out a trait type that sets a task local:

```swift
@Test(.withValue(123, for: $foo)) func f() { ... }
```

Or even if we don't make it a trait, can we make it easier to write a trait that sets a task local with minimal boilerplate?

Reply:
Agree that would be a very useful, built-in trait to have. Let's consider that separately, since it would ultimately leverage the API proposed here

Reply:
Doesn't need to be part of this proposal, but I think it does _affect_ this proposal. Might be worth coming up with an example of when you'd use this stuff other than to set a task local.

Reply:
Sure. Briefly: not all work before/after a test is simply mutating state in that process. It may involve writing to the filesystem, or interacting with other system-wide resources. (Clearly, parallelism would need to be considered in such scenarios.)

Reply:
Also: Users can still use this facility to mutate global or static state, as long as it complies with Swift concurrency rules

Reply:
Can we make sure to add this example to the discussion?

Reply:
Where, specifically, do you recommend we place this example code snippet?